meta-balena-common/recipes-core/systemd/systemd/65-resin-update-state.rules

```diff
@@ -7,6 +7,16 @@ ENV{ID_FS_LABEL_ENC}=="resin-*|balena-*", IMPORT{program}="/lib/udev/resin_updat
 ENV{ID_FS_LABEL_ENC}=="resin-*|balena-*", ENV{BALENA_FS_USE}=="?*", SYMLINK+="disk/by-state/$env{BALENA_FS_USE}"
 ENV{ID_FS_LABEL_ENC}=="resin-root*|balena-root*", ENV{BALENA_UPDATE_STATE}=="?*", SYMLINK+="disk/by-state/$env{BALENA_UPDATE_STATE}"
 
+IMPORT{cmdline}="flasher"
+ENV{flasher}=="?*", GOTO="balena_uuid_regen_end"
+# save the newly generated root UUID to override the original value passed on the cmdline
+ENV{BALENA_UPDATE_STATE}=="active", ENV{ID_FS_UUID}!="ba1eadef*|BDEF*", \
+        RUN+="/bin/sh -c \". /usr/libexec/os-helpers-fs && regenerate_uuid $devnode $env{ID_FS_TYPE} > /run/initramfs/bootparam_root \""
```

Comment:
The old logic was comparing the UUIDs with the `root=<uuid>` that the bootloader passed in the kernel command line to identify the root partition, and falling back to using labels. So the bootloader is the one that specifies what the booting disk is. 

What's the reason for the logic change?

Reply:
@alexgg This might be a little confusing, but I'll attempt to explain.

Currently, we have the [ `75-fsuuidsinit`](https://github.com/balena-os/meta-balena/blob/master/meta-balena-common/recipes-core/initrdscripts/files/fsuuidsinit) initramfs hook that waits for the by-state links (`resin-rootA` in particular) to be created. It would then check to see if the UUIDs had been regenerated, and would do so if necessary. If the UUID of the partition [matches](https://github.com/balena-os/meta-balena/blob/a84dc6fbc5967f67acf0d12f4fe5a9081a70cb63/meta-balena-common/recipes-core/initrdscripts/files/fsuuidsinit#L61) the UUID of the root passed on the kernel cmdline, we write the new UUID to `/run/initramfs/bootparam_root` after regenerating.

Before this we run udev, and have udev rules running in parallel while devices are initialized, including [`65-resin-update-state.rules`](https://github.com/balena-os/meta-balena/blob/master/meta-balena-common/recipes-core/systemd/systemd/65-resin-update-state.rules). One of these rules runs `resin_update_state_probe` for all balenaOS partitions. This script [already determines](https://github.com/balena-os/meta-balena/blob/a84dc6fbc5967f67acf0d12f4fe5a9081a70cb63/meta-balena-common/recipes-core/systemd/systemd/resin_update_state_probe#L52) if the partition is on the same drive as the root passed on the kernel cmdline, and adds an env var to the device's attributes for filtering, so we can simply add another rule below this to regenerate the UUIDs for these devices.

This is also why there are two similar rules, one for the root partition to write the new UUID to `/run`, and one for the remaining partitions.

I'm also not opposed to writing some tests for these behaviors, to ensure they remain working as intended.